File: src/LCR150.cong-shang-dao-xia-da-yin-er-cha-shu-ii-lcof.100312/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> decorateRecord(TreeNode root) {
        if(root == null){
            return List.of();
        }
        List<List<Integer>> result = new ArrayList<>();
        travel(List.of(root),result);
        return result;
    }
    
    private void travel(List<TreeNode> row, List<List<Integer>> result){
        List<Integer> vals = new ArrayList<>();
        List<TreeNode> next = new ArrayList<>();
        for(TreeNode node : row){
            vals.add(node.val);
            if(node.left != null){
                next.add(node.left);
            }
            if(node.right != null){
                next.add(node.right);
            }
        }
        result.add(vals);
        if(!next.isEmpty()){
            travel(next,result);
        }
        
    }
}

File: src/LCR150.cong-shang-dao-xia-da-yin-er-cha-shu-ii-lcof.100312/solution.ts
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor (val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val);
        this.left = (left === undefined ? null : left);
        this.right = (right === undefined ? null : right);
    }
}

function levelOrder (root: TreeNode | null): number[][] {
    if (root === null) {
        return [];
    }
    const result:number[][] = [];
    bfs([root, ], result);
    return result;
}

function bfs (levelNodes:TreeNode[], result:number[][]):void {
    const levelVals:number[] = [];
    const nextNodes:TreeNode[] = [];
    for (let i = 0; i < levelNodes.length; i++) {
        const node = levelNodes[i];
        levelVals.push(node.val);
        node.left !== null && nextNodes.push(node.left);
        node.right !== null && nextNodes.push(node.right);
    }
    result.push(levelVals);
    nextNodes.length > 0 && bfs(nextNodes, result);
}
